src/metabase/api/card.clj

```diff
@@ -348,22 +414,40 @@
    collection_id          (s/maybe su/IntGreaterThanZero)
    result_metadata        (s/maybe results-metadata/ResultsMetadata)
    metadata_checksum      (s/maybe su/NonBlankString)}
-  (let [card (api/write-check Card id)]
+  (let [card-before-update (api/write-check Card id)]
     ;; Do various permissions checks
-    (check-allowed-to-change-collection card collection_id)
-    (check-allowed-to-modify-query card dataset_query)
-    (check-allowed-to-unarchive card archived)
-    (check-allowed-to-change-embedding card enable_embedding embedding_params)
+    (check-allowed-to-change-collection card-before-update collection_id)
+    (check-allowed-to-modify-query card-before-update dataset_query)
+    (check-allowed-to-unarchive card-before-update archived)
+    (check-allowed-to-change-embedding card-before-update enable_embedding embedding_params)
     ;; make sure we have the correct `result_metadata`
-    (let [body (assoc body :result_metadata (result-metadata-for-updating card dataset_query result_metadata metadata_checksum))]
+    (let [body (assoc body :result_metadata (result-metadata-for-updating card-before-update dataset_query result_metadata metadata_checksum))]
       ;; ok, now save the Card
       (db/update! Card id
         ;; `collection_id` and `description` can be `nil` (in order to unset them). Other values should only be modified if they're passed in as non-nil
         (u/select-keys-when body
           :present #{:collection_id :description}
           :non-nil #{:dataset_query :display :name :visualization_settings :archived :enable_embedding :embedding_params :result_metadata})))
     ;; Fetch the updated Card from the DB
-    (let [card (Card id)]
+    (let [card   (Card id)
+          alerts (pulse/retrieve-alerts-for-card id)]
+
+      ;; If there are alerts, we need to check to ensure the card change doesn't invalidate the alert
+      (when (seq alerts)
+        (cond
+
+          (card-archived? card-before-update card)
+          (delete-alert-and-notify-archived! alerts)
+
+          (or (display-change-broke-alert? card-before-update card)
+              (goal-missing? card-before-update card)
+              (multiple-breakouts? card))
+          (delete-alert-and-notify-changed! alerts)
+
+          ;; The change doesn't invalidate the alert, do nothing
+          :else
+          nil))
+
```

Comment:
I think this endpoint has too much going on now. We should move this entire chunk of logic into a separate function and keep the endpoint itself as high-level as possible. I'd prefer a single call to something `(delete-alerts-if-needed! card)` in the endpoint and that function itself could fetch the alerts and do all of the checks there.